core/src/main/java/org/elasticsearch/action/search/RemoteClusterConnection.java

```diff
@@ -0,0 +1,492 @@
+            connectHandler.connect(ActionListener.wrap((x) -> fetchShardsInternal(searchRequest, indices, listener), listener::onFailure));
+        } else {
+            fetchShardsInternal(searchRequest, indices, listener);
+        }
+    }
+
+    private void fetchShardsInternal(SearchRequest searchRequest, List<String> indices,
+                                     final ActionListener<ClusterSearchShardsResponse> listener) {
+        final DiscoveryNode node = nodeSupplier.get();
+        ClusterSearchShardsRequest searchShardsRequest = new ClusterSearchShardsRequest(indices.toArray(new String[indices.size()]))
+            .indicesOptions(searchRequest.indicesOptions()).local(true).preference(searchRequest.preference())
+            .routing(searchRequest.routing());
+        transportService.sendRequest(node, ClusterSearchShardsAction.NAME, searchShardsRequest,
+            new TransportResponseHandler<ClusterSearchShardsResponse>() {
+
+                @Override
+                public ClusterSearchShardsResponse newInstance() {
+                    return new ClusterSearchShardsResponse();
+                }
+
+                @Override
+                public void handleResponse(ClusterSearchShardsResponse clusterSearchShardsResponse) {
+                    listener.onResponse(clusterSearchShardsResponse);
+                }
+
+                @Override
+                public void handleException(TransportException e) {
+                    listener.onFailure(e);
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SEARCH;
+                }
+            });
+    }
+
+    /**
+     * Returns a connection to the remote cluster. This connection might be a proxy connection that redirects internally to the
+     * given node.
+     */
+    Transport.Connection getConnection(DiscoveryNode remoteClusterNode) {
+        DiscoveryNode discoveryNode = nodeSupplier.get();
+        Transport.Connection connection = transportService.getConnection(discoveryNode);
+        return new Transport.Connection() {
+            @Override
+            public DiscoveryNode getNode() {
+                return remoteClusterNode;
+            }
+
+            @Override
+            public void sendRequest(long requestId, String action, TransportRequest request, TransportRequestOptions options)
+                throws IOException, TransportException {
+                connection.sendRequest(requestId, TransportActionProxy.getProxyAction(action),
+                    TransportActionProxy.wrapRequest(remoteClusterNode, request), options);
+            }
+
+            @Override
+            public void close() throws IOException {
+                assert false: "proxy connections must not be closed";
+            }
+        };
+    }
+
+    /**
+     * Returns the name of the remote cluster
+     */
+    public String getClusterName() {
```

Comment:
Similar to the comment I left below in the docs, this is not really the cluster name is it? I think we should make some effort to either clarify the terminology or make it the actual cluster name and validate it.